### PR TITLE
NeMo PPO

### PR DESCRIPTION
**Pull request by cat-state**
This PR adds `NeMoPPOTrainer`, a trainer for training with PPO NeMo Megatron models, supporting tensor parallelism and reference model offloading. 
20b: https://wandb.ai/uwu1/trlxnemo/runs/cg6ejr3m?workspace=user-uwu1

### CODE REVIEW

File: trlx/models/modeling_nemo_ilql.py [hunk 1]
```diff
@@ -783,4 +756,30 @@ def generate(
...
+
+        if activations_checkpointing_was_enabled:
+            self.activation_checkpointing_(False)
+
+        gen = super().generate(inputs, length_params, sampling_params)
+
+        if activations_checkpointing_was_enabled:
+            self.activation_checkpointing_(True)
+
+        if sp_was_enabled:
```
Comment (LouisCastricato) [inline trlx/models/modeling_nemo_ilql.py:773]: Why do you set it twice if sp was enabled? first to false then to true haha.

File: trlx/trainer/accelerate_ppo_trainer.py [hunk 2]
```diff
@@ -435,7 +437,7 @@ def make_experience(self, num_rollouts: int = 1024, iter_count: int = 0):  # noq
                 start = prompt_tensors.shape[1] - 1
 
             log_ratio = (logprobs - ref_logprobs) * attention_mask[:, :-1]
-            self.mean_kl = (log_ratio.exp() - 1 - log_ratio).mean().to(device)
+            self.mean_kl += (log_ratio.exp() - 1 - log_ratio).sum().to(device)
```
Comment (LouisCastricato) [inline trlx/trainer/accelerate_ppo_trainer.py:440]: why are we changing the accelerate implementation? Shouldn't this be its own PR?

File: trlx/trainer/nemo_ppo_trainer.py [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+        metric_fn: Optional[Callable[[List[str]], Any]] = None,
+        megatron_cfg: Optional[Union[str, dict]] = None,
+        pretrained_model: Optional[str] = None,
+        **kwargs,
+    ):
+        super().__init__(config, metric_fn=metric_fn, **kwargs)
+
+        if metric_fn is None:
+
+            def metric_fn(*args, **kwargs):
```
Comment (LouisCastricato) [inline trlx/trainer/nemo_ppo_trainer.py:45]: honestly we should print a warning if someone doesn't set the metric function.

File: trlx/trainer/nemo_ppo_trainer.py [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+            input_ids = batch["input_ids"].to(device)
+            lengths = lengths.to(device)
+            pad_to = lengths.max().item() + max_new_tokens
+
+            pad_to = ceil(pad_to / 8) * 8
+            input_ids = torch.nn.functional.pad(
+                input_ids, (0, pad_to - input_ids.shape[1]), value=self.tokenizer.pad_token_id
+            )
+
+            self.model.offload_reference_model()
```
Comment (LouisCastricato) [inline trlx/trainer/nemo_ppo_trainer.py:145]: Maybe a comment explaining why we need to offload every rollout?

**Review by LouisCastricato — COMMENTED** [review]
Will review more later... theres a lot of code to read 😅

File: trlx/trainer/nemo_ppo_trainer.py (hunk 3 above)
Comment (cat-state) [reply trlx/trainer/nemo_ppo_trainer.py:45]: for accelerate we automatically set it to the reward fn too

File: trlx/trainer/nemo_ppo_trainer.py (hunk 4 above)
Comment (cat-state) [reply trlx/trainer/nemo_ppo_trainer.py]: removed since its a noop, leftover code

File: trlx/models/modeling_nemo_ilql.py (hunk 1 above)
Comment (cat-state) [reply trlx/models/modeling_nemo_ilql.py:773]: disable if enabled and then re-enable if enabled

**Comment by cat-state** [comment]
theres a bit of cleanup with config handling that could be done - e.g setting nemo config fields from trlx config fully, adding default_nemo_config but probably best to do that in a subsequent PR to not further bloat this one

**Comment by cat-state** [comment]
Got pretrained format saving and loading working:
```python
    model = PPOGPT(ppo_config=ppo_config, cfg=megatron_cfg.model, trainer=trainer)
    model.load_from_pretrained(_20b_checkpoint_path)

    test = ["I don't know much about Hungarian underground"]
    test = [model.tokenizer.tokenizer.bos_token + t for t in test]

    print(model.generate(test, dict(max_length=40, min_length=0), dict(use_greedy=True))["sentences"])
    model.save_pretrained("infer_test_dir")
    model.load_from_pretrained("infer_test_dir")

    test = ["I don't know much about Hungarian underground"]
    test = [model.tokenizer.tokenizer.bos_token + t for t in test]

    print(model.generate(test, dict(max_length=40, min_length=0), dict(use_greedy=True))["sentences"])
```

```
2: ["I don't know much about Hungarian underground music, but I do know that the Hungarian government is a bunch of fascists.\n\nI don't know much about Hungarian underground music, but I do know that the Hungarian government is a bunch of"]
0: ["I don't know much about Hungarian underground music, but I do know that the Hungarian government is a bunch of fascists.\n\nI don't know much about Hungarian underground music, but I do know that the Hungarian government is a bunch of"]
1: ["I don't know much about Hungarian underground music, but I do know that the Hungarian government is a bunch of fascists.\n\nI don't know much about Hungarian underground music, but I do know that the Hungarian government is a bunch of"]
3: ["I don't know much about Hungarian underground music, but I do know that the Hungarian government is a bunch of fascists.\n\nI don't know much about Hungarian underground music, but I do know that the Hungarian government is a bunch of"]
1: Saving to infer_test_dir/mp_rank_01/model_weights.ckpt
0: Saving to infer_test_dir/mp_rank_00/model_weights.ckpt
2: Saving to infer_test_dir/mp_rank_02/model_weights.ckpt
3: Saving to infer_test_dir/mp_rank_03/model_weights.ckpt
0: Loading from infer_test_dir/mp_rank_00/model_weights.ckpt
1: Loading from infer_test_dir/mp_rank_01/model_weights.ckpt
2: Loading from infer_test_dir/mp_rank_02/model_weights.ckpt
3: Loading from infer_test_dir/mp_rank_03/model_weights.ckpt
0: Loaded from pretrained infer_test_dir/mp_rank_00/model_weights.ckpt
1: Loaded from pretrained infer_test_dir/mp_rank_01/model_weights.ckpt
2: Loaded from pretrained infer_test_dir/mp_rank_02/model_weights.ckpt
3: Loaded from pretrained infer_test_dir/mp_rank_03/model_weights.ckpt
0: ["I don't know much about Hungarian underground music, but I do know that the Hungarian government is a bunch of fascists.\n\nI don't know much about Hungarian underground music, but I do know that the Hungarian government is a bunch of"]
1: ["I don't know much about Hungarian underground music, but I do know that the Hungarian government is a bunch of fascists.\n\nI don't know much about Hungarian underground music, but I do know that the Hungarian government is a bunch of"]
2: ["I don't know much about Hungarian underground music, but I do know that the Hungarian government is a bunch of fascists.\n\nI don't know much about Hungarian underground music, but I do know that the Hungarian government is a bunch of"]
3: ["I don't know much about Hungarian underground music, but I do know that the Hungarian government is a bunch of fascists.\n\nI don't know much about Hungarian underground music, but I do know that the Hungarian government is a bunch of"]
```

File: trlx/trainer/nemo_ppo_trainer.py [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+    def add_prompt_pipeline(self, pipeline: PromptPipeline):
+        self.prompt_pipeline = pipeline
+
+    def make_experience(self, prompt_iterator: Iterator, num_rollouts: int = 1024, dp_world: int = 1):
+        ppo_rl_elements = []
+
+        device = torch.device("cuda")
+
+        if torch.distributed.get_rank() == 0:
+            tbar = tqdm(total=num_rollouts * dp_world, desc="Generating experience")
```
Comment (reciprocated) [inline trlx/trainer/nemo_ppo_trainer.py:125]: Nit:
```suggestion
            tbar = tqdm(total=num_rollouts * dp_world, desc="Generating experience", disable=torch.distributed.get_rank())
```
This should remove the need for checks in further `update`/`close` calls

Comment (cat-state) [reply trlx/trainer/nemo_ppo_trainer.py]: Cleaned up the tbar handling a bit

File: examples/nemo_ppo_sentiments.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+    config = default_config.evolve(
+        train=dict(
+            seq_length=2048,
+            batch_size=32,
+            epochs=100,
+            eval_interval=64,
+            trainer="NeMoPPOTrainer",
+            trainer_kwargs=dict(
+                pretrained_model=f"/mnt/nvme/home/uwu/nemo-megatron-gpt-{cfg_name}/",
```
Comment (maxreciprocate) [inline examples/nemo_ppo_sentiments.py:45]: maybe we should move these checkpoints to `/mnt/hdd`, otherwise they aren't accessible because of permissions

File: examples/nemo_ppo_inference.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    names = [os.path.basename(c) for c in checkpoints]
+    return set(names)
+
+
+def main(megatron_cfg_path, checkpoint_path):
+    ppo_config = trl_config.method
+
+    megatron_cfg = OmegaConf.load(megatron_cfg_path)
+    megatron_cfg.trainer.num_nodes = 1
+    megatron_cfg.trainer.devices = 4
```
Comment (maxreciprocate) [inline examples/nemo_ppo_inference.py:39]: For me inference fails with a single process, unless I manually set global batch size here, like so:
```python
megatron_cfg.model.global_batch_size = 64
```
But it also fails for me with multiprocessing, running `srun --label $PY -u /mnt/nvme/home/reciprocate/trlx/examples/nemo_ppo_inference.py configs/nemo_configs/megatron_1.3b.yaml nemo_1.3B_ppo_sentiments
` over `ntasks-per-node=8` with
```
4: │ ❱ 79 │   main(sys.argv[1], sys.argv[2])                                      │                                                                                                                                              [198/1868]
4: │   80                                                                         │
4: │                                                                              │
4: │ /mnt/nvme/home/reciprocate/trlx/examples/nemo_ppo_inference.py:61 in main    │
4: │                                                                              │
4: │   58 │   │   app_state.data_parallel_size,                                   │
4: │   59 │   │   app_state.pipeline_model_parallel_split_rank,                   │
4: │   60 │   │   app_state.virtual_pipeline_model_parallel_rank,                 │
4: │ ❱ 61 │   ) = fake_initialize_model_parallel(                                 │
4: │   62 │   │   world_size=app_state.model_parallel_size,                       │
4: │   63 │   │   rank=trainer.global_rank,                                       │
4: │   64 │   │   tensor_model_parallel_size_=megatron_cfg.model.tensor_model_par │
4: │                                                                              │
4: │ /mnt/nvme/home/reciprocate/miniconda3/envs/nemo/lib/python3.8/site-packages/ │
4: │ nemo/collections/nlp/modules/common/megatron/megatron_init.py:225 in         │
4: │ fake_initialize_model_parallel                                               │
4: │                                                                              │
4: │   222 │   │   │   │   data_parallel_group = list(ranks)                      │
4: │   223 │   │   │   │   logging.info(f'Rank {rank} has data parallel group: {d │
4: │   224 │                                                                      │
4: │ ❱ 225 │   data_parallel_rank = data_parallel_group.index(rank)               │
4: │   226 │   logging.info(f'All data parallel group ranks: {all_data_parallel_g │
4: │   227 │   logging.info(f'Ranks {rank} has data parallel rank: {data_parallel │
4: │   228                                                                        │
4: ╰──────────────────────────────────────────────────────────────────────────────╯
4: UnboundLocalError: local variable 'data_parallel_group' referenced before
4: assignment
srun: error: cw-prod-a100-cu117-40: tasks 1-3,5-7: Exited with exit code 1

Comment (cat-state) [reply examples/nemo_ppo_inference.py:39]: oh okay ill try to repro

Comment (cat-state) [reply examples/nemo_ppo_inference.py]: Should be fixed now. Was due to 1.3B being tp=1 pp=1 and the inference assumes num_gpus = tp*pp (since it supports only dp world = 1)

Comment (cat-state) [reply examples/nemo_ppo_inference.py]: not sure why the global batch size thing is happening but automatically setting it to mbs seems to work fine

File: examples/nemo_ppo_sentiments.py (hunk 6 above)
Comment (cat-state) [reply examples/nemo_ppo_sentiments.py]: moved

File: trlx/trainer/nemo_ppo_trainer.py [hunk 8]
```diff
@@ -0,0 +1,447 @@
...
+            query_tokens, response_tokens = zip(*query_responses)
+
+            query_tensors = [torch.tensor(t, device="cpu") for t in query_tokens]
+            response_tensors = [torch.tensor(t, device="cpu") for t in response_tokens]
+            logprobs = logprobs.cpu()
+            ref_logprobs = ref_logprobs.cpu()
+
+            masks = attention_mask.cpu()
+            log_ratio = (logprobs - ref_logprobs) * masks[:, :-1]
+            print(f"{log_ratio=}")
```
Comment (maxreciprocate) [inline trlx/trainer/nemo_ppo_trainer.py:247]: Nit: 
```suggestion
```

File: trlx/trainer/nemo_ppo_trainer.py [hunk 9]
```diff
@@ -0,0 +1,447 @@
...
+        from nemo.collections.nlp.modules.common.transformer.text_generation import (
+            LengthParam,
+        )
+
+        length = LengthParam(min_length=10, max_length=100)
+        generate = self.model.generate(["hello how are you"], length_params=length)
+        print(generate)
+        # self.model.save_pretrained("llama-nemo-7b-tp4")
+        # exit()
+        # import ipdb; ipdb.set_trace()
```
Comment (maxreciprocate) [inline trlx/trainer/nemo_ppo_trainer.py:373]: Nit:
```suggestion
```

File: trlx/trainer/nemo_ppo_trainer.py [hunk 10]
```diff
@@ -0,0 +1,447 @@
...
+            logprobs = logprobs.cpu()
+            ref_logprobs = ref_logprobs.cpu()
+
+            masks = attention_mask.cpu()
+            log_ratio = (logprobs - ref_logprobs) * masks[:, :-1]
+            print(f"{log_ratio=}")
+            for query_tensor, response_tensor, logps, vs, kl_penalty, score, start, mask in zip(
+                query_tensors, response_tensors, logprobs, values, log_ratio, scores, lengths, masks
+            ):
+                start = start
```
Comment (maxreciprocate) [inline trlx/trainer/nemo_ppo_trainer.py:251]: Nit:
```suggestion
```

**Review by maxreciprocate — APPROVED** [review]
Thanks Aman for your thorough work on this!
LGTM

File: trlx/trainer/nemo_ppo_trainer.py (hunk 9 above)
Comment (cat-state) [reply trlx/trainer/nemo_ppo_trainer.py:373]: Oops, accidently backported these from llama, will remove
